common/workunit/referencedfilelist.cpp

```diff
@@ -474,13 +488,213 @@ void ReferencedFile::resolve(const char *dstCluster, const char *srcCluster, IUs
+
+    StringBuffer destFolder, destTitle, defaultFolder, defaultReplicateFolder;
+    StringBuffer srcNodeGroup;
+    StringBuffer destNodeGroup(destNodeGroupIn);
+    if(destNodeGroup.isEmpty())
+    {
+        getNodeGroupFromLFN(user, sourceLogicalName, destNodeGroup);
+        DBGLOG("Destination node group not specified, using source node group %s", destNodeGroup.str());
+    }
+
+    CDfsLogicalFileName lfn;
+    parseLogicalPath(destLogicalName, destNodeGroup.str(), NULL, destFolder, destTitle, defaultFolder, defaultReplicateFolder);
+
+    StringBuffer fileMask;
+    constructFileMask(destTitle.str(), fileMask);
+
+    CDfsLogicalFileName logicalName;
+    logicalName.set(sourceLogicalName);
+    if (!isEmptyString(srcDali))
+    {
+        SocketEndpoint ep(srcDali);
+        if (ep.isNull())
+            throw MakeStringException(-1, "ReferencedFile Copy %s: cannot resolve SourceDali network IP from %s.", sourceLogicalName, srcDali);
+
+        logicalName.setForeign(ep,false);
+    }
+
+    Owned<IDistributedFile> file = queryDistributedFileDirectory().lookup(logicalName, user, false, false, false, nullptr, defaultPrivilegedUser);
+    if (!file)
+        throw MakeStringException(-1, "ReferencedFile failed to find file: %s", logicalName.get());
+
+    if (supercopy)
+    {
+        if (!file->querySuperFile())
+            supercopy = false;
+    }
+    else if (file->querySuperFile() && (file->querySuperFile()->numSubFiles() > 1) && isFileKey(file))
+        supercopy = true;
+
+    StringBuffer dfuQueueName(dfu_queue);
+    if (dfuQueueName.isEmpty())
+        getDefaultDFUName(dfuQueueName);
+
+    Owned<IDFUWorkUnitFactory> factory = getDFUWorkUnitFactory();
+    Owned<IDFUWorkUnit> wu = factory->createPublisherSubTask(publisherWuid);
+    wu->setJobName(destLogicalName);
+    wu->setQueue(dfuQueueName);
+
+    if (user)
+    {
+        StringBuffer username, password;
+        wu->setUser(user->getUserName(username).str());
+        wu->setPassword(user->getPassword(password).str());
+    }
+
+    if(!destNodeGroup.isEmpty())
+        wu->setClusterName(destNodeGroup.str());
+    if (supercopy)
+        wu->setCommand(DFUcmd_supercopy);
+    else
+        wu->setCommand(DFUcmd_copy);
+
+    IDFUfileSpec *wuFSpecSource = wu->queryUpdateSource();
+    IDFUfileSpec *wuFSpecDest = wu->queryUpdateDestination();
+    IDFUoptions *wuOptions = wu->queryUpdateOptions();
+    wuFSpecSource->setLogicalName(sourceLogicalName);
+    if (!isEmptyString(srcDali))
+    {
+        SocketEndpoint ep(srcDali);
+        wuFSpecSource->setForeignDali(ep);
+
+        StringBuffer srcUserName;
+        user->getUserName(srcUserName);
+        if(!srcUserName.isEmpty())
+        {
+            StringBuffer srcPassWord;
+            user->getPassword(srcPassWord);
+            wuFSpecSource->setForeignUser(srcUserName, srcPassWord);
+        }
+    }
+    wuFSpecDest->setLogicalName(destLogicalName);
+    wuFSpecDest->setFileMask(fileMask.str());
```

Comment:
@jakesmith does code need to set file mask, or only if it differs from the default?